WDL/runtime/backend/singularity.py

```diff
@@ -49,11 +67,16 @@ def cli_exe(self) -> List[str]:
     def _pull_invocation(self, logger: logging.Logger, cleanup: ExitStack) -> Tuple[str, List[str]]:
         image, invocation = super()._pull_invocation(logger, cleanup)
         docker_uri = "docker://" + image
-
-        # The docker image layers are cached in SINGULARITY_CACHEDIR, so we don't need to keep the
-        # *.sif
-        pulldir = cleanup.enter_context(tempfile.TemporaryDirectory(prefix="miniwdl_sif_"))
-        return (docker_uri, self.cli_exe + ["pull", "--dir", pulldir, docker_uri])
+        if self.image_cache_dir is not None:
+            image_name = docker_uri.replace("/", "_").replace(":", "_")
+            image_path = os.path.join(self.image_cache_dir, image_name)
+            if not os.path.exists(image_path):
+                return (image_path, self.cli_exe + ["pull", image_path, docker_uri])
+            # If path already exists test the image
+            return (image_path, self.cli_exe + ["exec", image_path, "true"])
```

Comment:
I think we can modify the SubprocessBase class to allow us to `return (image_path, None)` here to indicate that we don't actually need to pull anything. (But it still needs to acquire the "pull lock" for a brief instance, for the aforementioned reason)

IIRC, if the image is large then Singularity takes some time to start up a container even just to run `true`.

Reply:
That is indeed a much better option. Will implement.